mipengine/node/monetdb_interface/monet_db_facade.py

```diff
@@ -212,3 +218,32 @@ def _get_lock_timeout_error_msg(db_execution_dto, lock_timeout):
         lock was not acquired during
         {lock_timeout=}
     """
+
+
+def _create_fault_tolerant_query(query: str):
+    """
+    This is a query optimization method to protect from the following edge case:
+    1) A udf starts running allocating memory,
+    2) a table creation query starts running,
+    3) the udf allocates more memory than monetdb can provide,
+    4) the container memory limit kills monetdb,
+    5) the table creation query returns with BrokenPipeError("Server closed the connection"),
+    6) when the monet_db_facade tries to rerun the failed query it receives a "Table already exists error"
+        because the table was created even though the connection was severed.
+
+    The solution to this is a suffix on the "CREATE TABLE" queries with "IF NOT EXISTS" to cover this scenario.
+    """
+    fault_tolerant_query = query
+    fault_tolerant_query = fault_tolerant_query.replace(
+        "CREATE TABLE", "CREATE TABLE IF NOT EXISTS"
+    )
+    fault_tolerant_query = fault_tolerant_query.replace(
+        "CREATE REMOTE TABLE", "CREATE REMOTE TABLE IF NOT EXISTS"
+    )
+    fault_tolerant_query = fault_tolerant_query.replace(
+        "CREATE MERGE TABLE", "CREATE MERGE TABLE IF NOT EXISTS"
+    )
+    fault_tolerant_query = fault_tolerant_query.replace(
+        "CREATE VIEW", "CREATE OR REPLACE VIEW"
+    )
```

Comment:
Why are query strings modified here instead of being fixed at the source, where they are created? 

Mutating a value that was created earlier can lead to difficult to understand behavior. One can no longer understand the system's behavior just by looking at the function that creates the query. Instead, one has to follow the query until it hits the DB.

Reply:
This approach was the initial fix but the problem was that this is a very specific fix for a very specific scenario. 

I don't want to pollute 5 different places (udfgenerator, tables, remote tables, merge tables, views) with each one containing part of the fix. It would be hard to maintain, for example if we add a new create query.

Also, the queries are created in a different place, but at that specific place there is no reason for the "IF NOT EXISTS" to be part of the query, it's not part of the use case. It is added in the "monetdb_facade" module since this fix is caused due to the db behavior so it's also solved at that layer.